### ISSUE TITLE
Autotune cannot survive a candidate that faults the device: the tuner panics, and no fallback design can recover

### ISSUE THREAD

**Issue by kafonek** [kafonek/cubecl#8]
When every candidate in an autotune set fails, the tuner panics instead of degrading. Worse, one of the failure modes we hit — a candidate that triggers `CUDA_ERROR_ILLEGAL_ADDRESS` while being profiled — **cannot be recovered from by any in-process fallback**, because the fault poisons the CUDA context that a fallback would have to run in.

That distinction matters for the design. "Catch the error and try the next candidate" is a reasonable goal for a candidate that returns a setup error. It is unreachable for a candidate that faults the device. Recovery has to happen before launch, not after.

This is the failure-semantics half of a crash we filed separately as #3, which covers the out-of-bounds access itself.

### The deterministic case

1x NVIDIA H100 80GB HBM3, Linux container, CUDA 12.6. `burn 0.22.0-pre.2`, `cubecl 0.11.0-pre.2`, `cubek 0.3.0-pre.2`, CUDA backend. A small Rust binary wrapping burn's public API that runs a convolution operator benchmark.

```
CUBECL_AUTOTUNE_SHORT_CIRCUIT=false CUBECL_DEBUG_LOG=stdout \
  <bench> --backend cuda --dtype bf16 --backward --iterations 5
```

`conv2d`, bf16, under autodiff, shape `[32, 256, 64, 64]`, `Conv2dConfig::new([256, 256], [3, 3])` with `PaddingConfig2d::Same`.

**All four accelerated convolution-forward candidates** — `simple_async_cmma`, `simple_async_mma`, `simple_tma_cmma`, `simple_tma_mma` — hit `CUDA_ERROR_ILLEGAL_ADDRESS` inside `conv_forward_nhwc` during profiling, on the first pass, reproducibly. The tuner then panics:

```
cubecl_runtime::tune::schedule::Schedule::run_plan
  called `Result::unwrap()` on an `Err` value
```

Two observations about why this was hard to find:

- **With short-circuiting on (the default) it is intermittent.** `AutotuneConfig::disable_short_circuit` is `#[serde(default)]` and therefore `false`, and `BenchConfig::short_circuit_samples` is 2, so candidates are eliminated after two samples against the current leader. Which candidates fully execute varies run to run with timing noise, and a candidate with an out-of-bounds access only faults on runs where it is actually reached. An identical rerun of the original failing job completed cleanly. That is the whole explanation for the intermittency, and it took disabling short-circuiting to see it.
- **The crash surfaces at the wrong place.** Work is enqueued asynchronously, so an illegal access is reported by whichever call next touches the stream. In a training loop that is the next iteration's forward, which is why the first stack trace we collected pointed at `Conv2d::forward` and not at the tuner.

### Why a candidate-fallback cannot fix this case

`CUDA_ERROR_ILLEGAL_ADDRESS` is sticky. Once a kernel indexes outside its allocation, the context is unusable and every subsequent call on it returns the same error until the context is destroyed. A fallback that catches the error and executes the next candidate would execute it in a dead context and get the same error back.

There is a second, independent obstacle that has already been raised in this ecosystem's design discussions: a fallback would have to hand the *real* inputs to a second candidate, and autotune must never clone them, because an in-place operation would write into the copy. That objection is correct and it applies to every fallback design, not just ours.

Put together, the two constraints point the same way: **prevention beats recovery**. Concretely,

1. **Validate before launching.** Where a candidate's preconditions are checkable — layout, alignment, shared-memory budget, stride requirements — check them at selection time and give the candidate a "never" priority when they do not hold, rather than discovering it at setup after autotune has committed. This is the approach another reporter independently arrived at for an alignment-related matmul failure in this same version, and it is the one that generalises.
2. **Where a candidate must actually run to be judged**, accept that a device fault is terminal for the process and make it *loud and specific* rather than a bare `unwrap`. A panic that names the candidate, the autotune key and the problem definition costs nothing and is the difference between a one-hour diagnosis and a week.
3. **Do not let the cache-hit path assert.** `LocalTuner::execute` (`cubecl-runtime-0.11.0-pre.2/src/tune/local.rs`) ends the fast path with

   ```rust
   return operations
       .fastest(fastest_index)
       .execute(inputs)
       .expect("Should run when selected by autotune.");
   ```

   A winner recorded in a persistent cache can be replayed against a problem whose real layout differs from the one it was tuned on, and this turns that into a panic on a path where re-tuning is still possible.

### The knob that makes any of this reproducible

`CUBECL_AUTOTUNE_SHORT_CIRCUIT=false` should be the first thing anyone reaches for when an autotune-adjacent crash is intermittent. It forces every candidate to be benchmarked, so a faulty candidate is reached on every run instead of occasionally. It converted this bug from "did not reproduce on a rerun" into "fires on the first pass, every time". It is discoverable only by reading `cubecl-runtime-0.11.0-pre.2/src/config/autotune.rs` and `src/config/base.rs`, where the environment variable inverts into `disable_short_circuit`. Worth documenting.

### Not verified

- I did not run `compute-sanitizer`, which with short-circuiting disabled should now name the exact kernel and address on the first run. That is the obvious next step and it belongs to #3 rather than here.
- I did not check whether the same panic path is reached when a candidate returns an ordinary setup error rather than faulting the device, so I cannot say what fraction of real failures are recoverable in principle.
- I did not test whether destroying and recreating the CUDA context after a fault is feasible inside the runtime. I assume it is not worth it, but I did not investigate.
- I did not capture which candidate wins for the convolution gradients — the run crashed on the forward keys before reaching them.
